crates/consensus/beacon/src/engine/mod.rs

```diff
@@ -253,11 +263,34 @@ where
             metrics: Metrics::default(),
         };
 
-        if let Some(target) = target {
+        let maybe_pipeline_target = match target {
+            // Provided target always takes precedence.
+            target @ Some(_) => target,
+            None => {
+                // If no target was provided, check if the stages are congruent - check if the
+                // checkpoint of the last stage matches the checkpoint of the first.
+                let headers_stage_checkpoint = this
+                    .blockchain
+                    .get_stage_checkpoint(StageId::Headers)?
+                    .unwrap_or_default()
+                    .block_number;
+                let finish_stage_checkpoint = this
+                    .blockchain
+                    .get_stage_checkpoint(StageId::Finish)?
+                    .unwrap_or_default()
+                    .block_number;
+                if finish_stage_checkpoint != headers_stage_checkpoint {
```

Comment:
not sure if it's a good idea, but even though it is unlikely that we add some stage before `Headers`, it still would be nice to refer to some ordered list like so:
```suggestion
                let first_stage_checkpoint = this
                    .blockchain
                    .get_stage_checkpoint(*StageId::ALL.first().unwrap())?
                    .unwrap_or_default()
                    .block_number;
                let last_stage_checkpoint = this
                    .blockchain
                    .get_stage_checkpoint(*StageId::ALL.last().unwrap())?
                    .unwrap_or_default()
                    .block_number;
                if first_stage_checkpoint != last_stage_checkpoint {
```